### PR TITLE
Jenkins: Allow fetching new tars for multi-job tests

### LINKED ISSUE #10850: DNS broken after GKE upgrade
The Jenkins job for GKE upgrade, step 3 (e2e) has been failing since the 4:40pm (PST) run yesterday because of the following DNS problems:

```
DNS should provide DNS for the cluster

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/dns.go:226
Unexpected number of pods (0) matches the label selector k8s-app=kube-dns,kubernetes.io/cluster-service=true


DNS should provide DNS for services

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/dns.go:311
Unexpected number of pods (0) matches the label selector k8s-app=kube-dns,kubernetes.io/cluster-service=true
```

This is **after** the cluster is upgraded.

### PR DESCRIPTION
**Pull request by mbforbes**
This will allow us to do "true" upgrade testing by setting `JENKINS_FORCE_GET_TARS=y` in multi-step jobs where we want to retrieve a new binary.  This enables the following scenarios which currently cannot be exercised by Jenkins:
#### Use "new" code to do a full upgrade and run e2es

_This is how customers will likely do an upgrade. Also, it let's us correctly run e2es at the newer version, which fixes #10850 for our CI builds. Also, it let's us pin tests at release transitions, such as validating a 0.19.x&rarr;0.21.y upgrade._
1. Create cluster at version X (uses code at version X)
2. Upgrade cluster (master + nodes) to version Y (uses code at version Y) **[this PR enables X&rarr;Y switch]**
3. Run e2e tests at version Y (uses code at version Y)
#### Use "new" code to upgrade master; use "old" code to run e2es

_We can use this to ensure that upgrading only the master keeps the functionality at the old version of the cluster (as the nodes are still at the old version). This will mean we will still catch issues such as #10850, which was a true breaking behavioral change._
1. Create cluster at version X (uses code at version X)
2. Upgrade _only the master_ to version Y (uses code at version Y) **[this PR enables X&rarr;Y switch]**
3. Run e2e tests at version X (uses code at version X) **[this PR enables Y&rarr;X switch]**

### CODE REVIEW

**Comment by mbforbes** [comment]
I've been developing this against the GKE upgrade job, and the code in this PR is in flight on those jobs. The first two steps are passing (which includes the version change); step 3 is currently running.

**Comment by mbforbes** [comment]
Risk: low (doesn't affect any Jenkins job by default; will affect upgrade jobs once those are enabled with the new variable)
Impact: low (for Kubernetes), high (for testing releases—this is vital for upgrade testing!)

**Comment by mbforbes** [comment]
+cc @ixdy @zmerlynn

File: hack/jenkins/e2e.sh [hunk 1]
```diff
@@ -231,13 +231,19 @@ echo "Test Environment:"
 printenv | sort
 echo "--------------------------------------------------------------------------------"
 
-if [[ "${E2E_UP,,}" == "true" ]]; then
+# We get the kuberntes tarballs on either cluster creation or when we want to
+# replace existing ones in a multi-step job (e.g. a cluster upgrade).
+if [[ "${E2E_UP,,}" == "true" || ${JENKINS_FORCE_GET_TARS:-} =~ ^[yY]$ ]]; then
```
Comment (roberthbailey) [inline hack/jenkins/e2e.sh]: nit: quote the second var on this line

Comment (mbforbes) [inline hack/jenkins/e2e.sh]: We use a `=~` comparison about 5 other times in this file and never quote so I copied that style. Do you want me to go and quote all of them?

Comment (roberthbailey) [inline hack/jenkins/e2e.sh]: At least fix the newly introduced variables. Up to you if you want to fix the rest of the file as well.

Comment (mbforbes) [inline hack/jenkins/e2e.sh]: Done.

**Comment by mbforbes** [comment]
Thanks for the comment—nit addressed.

Ping for any reviewers? We've got a bunch of upgrade testing riding on this.

**Comment by zmerlynn** [comment]
Looking..

File: hack/jenkins/e2e.sh [hunk 2]
```diff
@@ -231,13 +231,19 @@ echo "Test Environment:"
...
     if [[ ${KUBE_RUN_FROM_OUTPUT:-} =~ ^[yY]$ ]]; then
         echo "Found KUBE_RUN_FROM_OUTPUT=y; will use binaries from _output"
         cp _output/release-tars/kubernetes*.tar.gz .
     else
         echo "Pulling binaries from GCS"
-        if [[ $(find . | wc -l) != 1 ]]; then
+        # In a multi-step job, clean up just the kubernetes build files.
+        # Otherwise, we want a completely empty directory.
+        if [[ "${JENKINS_FORCE_GET_TARS:-}" =~ ^[yY]$ ]]; then
+            rm -rf kubernetes*
```
Comment (zmerlynn) [inline hack/jenkins/e2e.sh]: Is this just trying to avoid destroying the test results?

Comment (mbforbes) [inline hack/jenkins/e2e.sh]: When I removed everything e2e.go vomited a giant Jenkins HTML page at the beginning of the tests when it tried to detect the version, so I think we need to keep `.kube` around or something?

I guess you could also look at this the other way, which is that the only thing I'm changing from the current setup is removing the tarballs and the directory they're splatted into.

Comment (zmerlynn) [inline hack/jenkins/e2e.sh]: Oh, `.kube`. Details.

**Comment by zmerlynn** [comment]
LGTM.

**Comment by zmerlynn** [comment]
@ixdy - If you get a chance to review, it'd be good, since we're pretending @quinton-hoole is out. But I'd like to get this merged ASAP.

**Comment by mbforbes** [comment]
As far as I know Jenkins isn't running this file (I think for PR builder it uses the e2e script at head rather than the one one the PR... possible future feature?) so the check could be nothing but noise. However, if you'd still like me to kick, I can.

**Comment by zmerlynn** [comment]
@mbforbes: It's actually running the `e2e.sh` at the checkout.

**Comment by zmerlynn** [comment]
@k8s-bot test this

@mbforbes: but you hit a flake

**Comment by roberthbailey** [comment]
lgtm.

**Comment by mbforbes** [comment]
@k8s-bot test this

File: hack/jenkins/e2e.sh [hunk 3]
```diff
@@ -231,13 +231,19 @@ echo "Test Environment:"
 printenv | sort
 echo "--------------------------------------------------------------------------------"
 
-if [[ "${E2E_UP,,}" == "true" ]]; then
+# We get the kuberntes tarballs on either cluster creation or when we want to
```
Comment (ixdy) [inline hack/jenkins/e2e.sh]: nit: kubernetes

**Comment by ixdy** [comment]
besides minor comment typo LGTM

**Comment by mbforbes** [comment]
I will add the `e` and start e2e and shippable again and then ping for an `ok-to-merge` :-)

**Comment by mbforbes** [comment]
Hit Jenkins flake and unsure what happened with shippable. Kicking...

**Comment by mbforbes** [comment]
@k8s-bot test this

**Comment by ixdy** [comment]
Pretty sure Shippable is just confused. This commit in no way affects it, anyway.

**Comment by vmarmol** [comment]
FYI: Waiting on Jenkins stability before merging.

**Comment by mbforbes** [comment]
Jenkins is green

**Comment by vmarmol** [comment]
Waiting on #11014 to be sorted out

**Comment by mbforbes** [comment]
Kicking shippable

**Pull request merged**